File: leetcode/leetcode3312.py
#https://leetcode.cn/problems/sorted-gcd-pair-queries/?envType=daily-question&envId=2026-07-17
from bisect import bisect_right
from itertools import accumulate
from typing import List
class Solution:
    def gcdValues(self, nums: List[int], queries: List[int]) -> List[int]:
        max_m=max(nums)
        cnt_x=[0]*(max_m+1)
        cnt_gcd=[0]*(max_m+1)
        for num in nums:
            cnt_x[num]+=1
        for i in range(max_m,0,-1):
            cnt_c=0
            for j in range(i,max_m+1,i):
                cnt_c+=cnt_x[j]
                cnt_gcd[i]-=cnt_gcd[j]
            cnt_gcd[i]+=(cnt_c*(cnt_c-1))/2
        s= list(accumulate(cnt_gcd))
        return [bisect_right(s,q) for q in queries]

File: leetcode/leetcode3312.cpp
#include<vector>
#include<algorithm>
using namespace std;
class Solution {
public:
    vector<int> gcdValues(vector<int>& nums, vector<long long>& queries) {
        int m=*max_element(nums.begin(),nums.end());
        vector<int> cnt_num(m+1);
        for(int &num:nums){
            cnt_num[num]++;
        }
        vector<long long> cnt_gcd(m+1);
        for(int i=m;i>0;i--){
            int c=0;
            for(int j=i;j<=m;j+=i){
                c+=cnt_num[j];
                cnt_gcd[i]-=cnt_gcd[j];
            }
            cnt_gcd[i]+= (long long)c * (c - 1) / 2; 
        }
        vector<long long> pre_sum(m+1);
        pre_sum[0]=0;
        for(int i=1;i<=m;i++){
            pre_sum[i]=cnt_gcd[i]+pre_sum[i-1];
        }
        vector<int> ans;
        for(long long &q:queries){
            ans.push_back(upper_bound(pre_sum.begin(),pre_sum.end(),q)-pre_sum.begin());
        }
        return ans;
    }
};